File: skills/todolist-md-clawdbot/scripts/apply_filtered_suggestions.mjs
#!/usr/bin/env node
import fs from 'fs'
import crypto from 'crypto'
import {request} from 'undici'
import {fileURLToPath} from 'url'
import path from 'path'
import {spawnSync} from 'child_process'

const __dirname = path.dirname(fileURLToPath(import.meta.url))
const SUGGESTIONS_PATH = path.resolve(process.cwd(), 'outputs/todolist-md/llm_suggestions_apply.json')
const BACKUPS_DIR = path.resolve(process.cwd(), 'outputs/todolist-md/backups')
const FILTER_RECORDS_DIR = path.resolve(process.cwd(), 'outputs/todolist-md/filter_records')
const THRESHOLD = parseFloat(process.env.SUGGEST_SIM_THRESHOLD || '0.85')
const ACTION_VERBS = (process.env.SUGGEST_ACTION_VERBS || 'create,implement,open,contact,draft,request,schedule,setup,configure,test,verify,follow').split(',')

function normalize(s){
  return (s||'').toLowerCase().replace(/[^a-z0-9 ]+/g,' ').replace(/\s+/g,' ').trim()
}
function similarity(a,b){
  // simple SequenceMatcher-like ratio via longest common subsequence approximated by token overlap
  a = normalize(a).split(' ')
  b = normalize(b).split(' ')
  if(!a.length || !b.length) return 0
  const setA = new Set(a)
  let common=0
  for(const t of b) if(setA.has(t)) common++
  return common / Math.max(a.length,b.length)
}
function isActionable(title){
  const t = normalize(title)
  for(const v of ACTION_VERBS){
    if(t.startsWith(v+' ') || t.indexOf(' '+v+' ')>=0) return true
  }
  return false
}

async function getAccessToken(){
  const cs = JSON.parse(fs.readFileSync('/home/openclaw/.openclaw/credentials/gog-client-secret.json','utf8'))
  const client_id = cs.installed?.client_id || cs.web?.client_id
  const client_secret = cs.installed?.client_secret || cs.web?.client_secret
  const rt = JSON.parse(fs.readFileSync('/home/openclaw/.openclaw/.secrets/todolist_drive_oauth.json','utf8')).refresh_token
  const body = new URLSearchParams({client_id,client_secret,refresh_token:rt,grant_type:'refresh_token'})
  const r = await request('https://oauth2.googleapis.com/token', { method: 'POST', body: body.toString(), headers: {'content-type':'application/x-www-form-urlencoded'} })
  const j = await r.body.json()
  return j.access_token
}

async function main(){
  if(!fs.existsSync(SUGGESTIONS_PATH)){
    console.error('suggestions missing:',SUGGESTIONS_PATH)
    process.exit(1)
  }
  const sugg = JSON.parse(fs.readFileSync(SUGGESTIONS_PATH,'utf8'))
  let items = []
  let fileId = null
  if(sugg.fileSuggestions && sugg.fileSuggestions[0]){
    const fsug = sugg.fileSuggestions[0]
    items = fsug.suggestions || fsug.filtered_and_normalized_suggestions || []
    fileId = fsug.fileId || fsug.fileId
  } else if(sugg.suggestions || sugg.filtered_and_normalized_suggestions){
    items = sugg.suggestions || sugg.filtered_and_normalized_suggestions || []
    fileId = sugg.fileId || (items[0] && items[0].fileId) || null
  } else if(sugg.fileId && sugg.suggestions){
    items = sugg.suggestions; fileId = sugg.fileId
  } else {
    console.error('unrecognized suggestions schema')
    process.exit(1)
  }
  const access = await getAccessToken()
  const headers = {authorization:`Bearer ${access}`}
  const metaR = await request(`https://www.googleapis.com/drive/v3/files/${fileId}?fields=id,name,mimeType,modifiedTime,headRevisionId`,{headers})
  const meta = await metaR.body.json()
  const dl = await request(`https://www.googleapis.com/drive/v3/files/${fileId}?alt=media`,{headers})
  const content = await dl.body.text()
  const lines = content.split(/\r?\n/)
  fs.mkdirSync(BACKUPS_DIR,{recursive:true})
  const bkname = path.join(BACKUPS_DIR, `${fileId}.${new Date().toISOString().replace(/[:.]/g,'')}.md`)
  fs.writeFileSync(bkname, content)
  let newLines = [...lines]
  let offset=0
  const records=[]
  let inserted=0
  for(const it of items){
    const orig = it.original
    const suggested = it.suggested_text || ''
    const title = suggested.replace(/^\- \[ \]\s*/,'').split('\n')[0]
    const sim = similarity(orig,title)
    const actionable = isActionable(title)
    let decision='insert'
    if(sim>=THRESHOLD) decision='skip_similar'
    else if(!actionable) decision='skip_not_actionable'
    records.push({original:orig,title,similarity:sim,actionable,decision})
    if(decision!=='insert') continue
    // find best match
    let bestIdx=null,bestScore=0
    for(let idx=0; idx<lines.length; idx++){
      const score = similarity(orig, lines[idx])
      if(score>bestScore){bestScore=score; bestIdx=idx}
    }
    if(bestIdx==null || bestScore<0.2) continue
    const short = title
    const humanNote = `  > <!-- bot: note --> ${short}`
    const subId = 'sub:'+crypto.randomBytes(4).toString('hex')
    const markerObj = {id:subId, original_line:orig, title:short, estimate_hours:null, assignee:null, createdAtUtc:new Date().toISOString(), source_model:'gpt-5-mini', status:'suggested'}
    const machineComment = `<!-- bot: subtask ${JSON.stringify(markerObj)} -->`
    const idx = bestIdx+1+offset
    const window = newLines.slice(Math.max(0,idx-2), Math.min(newLines.length, idx+3)).join('\n')
    if(window.includes(humanNote) || window.includes(machineComment)) continue
    newLines.splice(idx,0,humanNote,machineComment)
    offset+=2
    inserted++
  }
  // update last_review
  const lastLine = `<!-- bot: last_review --> ${new Date().toISOString()} model=gpt-5-mini hash=${sugg.generatedAtUtc||''}`
  if(newLines.length && newLines[0].startsWith('<!-- bot: last_review')) newLines[0]=lastLine
  else newLines.unshift(lastLine)
  const newContent = newLines.join('\n') + (content.endsWith('\n')? '\n':'')
  // upload
  const up = await request(`https://www.googleapis.com/upload/drive/v3/files/${fileId}?uploadType=media&supportsAllDrives=true`,{method:'PATCH', body:newContent, headers:{...headers,'Content-Type':meta.mimeType||'text/markdown'}})
  await up.body.text()
  fs.mkdirSync(FILTER_RECORDS_DIR,{recursive:true})
  const recfn = path.join(FILTER_RECORDS_DIR, `${fileId}.${new Date().toISOString().replace(/[:.]/g,'')}.json`)
  fs.writeFileSync(recfn, JSON.stringify(records,null,2))
  console.log('Backup',bkname)
  console.log('Inserted',inserted)
  console.log('Records ->',recfn)
}

main().catch(e=>{console.error(e); process.exit(1)})

File: skills/todolist-md-clawdbot/scripts/apply_filtered_suggestions.py
#!/usr/bin/env python3
"""
apply_filtered_suggestions.py
Post-process llm_suggestions_apply.json to avoid inserting near-duplicate suggestions
and apply as <!-- bot: subtask ... --> markers into target Drive files.

Behavior:
- Loads outputs/todolist-md/llm_suggestions_apply.json
- For each suggestion, compute similarity between original line and suggested title.
- Skip creating subtask marker if similarity >= 0.85 (configurable) -> instead record as no_change.
- Only insert subtasks for items that pass threshold and are actionable (start with verb or contain verbs like 'create','implement','open','contact').
- Backup file before write and write markers inline under matched lines.

Usage:
  python3 apply_filtered_suggestions.py --file vyond.md --apply

"""
import json,sys,os,re,argparse,uuid,datetime
from difflib import SequenceMatcher
import urllib.request,urllib.parse

THRESHOLD=0.85
ACTION_VERBS=('create','implement','open','contact','draft','request','schedule','setup','configure','test','verify','follow')

def normalize(s):
    return re.sub(r'\s+',' ', re.sub(r'[^a-z0-9 ]',' ', (s or '').lower())).strip()


def is_actionable(title):
    t=normalize(title)
    for v in ACTION_VERBS:
        if t.startswith(v+' ') or (' '+v+' ' in (' '+t+' ')):
            return True
    return False


def get_access_token():
    cs=json.load(open('/home/openclaw/.openclaw/credentials/gog-client-secret.json'))
    client_id=cs.get('installed',{}).get('client_id') or cs.get('web',{}).get('client_id')
    client_secret=cs.get('installed',{}).get('client_secret') or cs.get('web',{}).get('client_secret')
    rt=json.load(open('/home/openclaw/.openclaw/.secrets/todolist_drive_oauth.json')).get('refresh_token')
    data=urllib.parse.urlencode({'client_id':client_id,'client_secret':client_secret,'refresh_token':rt,'grant_type':'refresh_token'}).encode()
    req=urllib.request.Request('https://oauth2.googleapis.com/token',data=data)
    with urllib.request.urlopen(req) as resp:
        tok=json.load(resp)
    return tok.get('access_token')


def main():
    p=argparse.ArgumentParser()
    p.add_argument('--file',required=True)
    p.add_argument('--apply',action='store_true')
    p.add_argument('--threshold',type=float,default=THRESHOLD)
    args=p.parse_args()

    suggestions_path='outputs/todolist-md/llm_suggestions_apply.json'
    if not os.path.exists(suggestions_path):
        print('No suggestions file found at',suggestions_path); sys.exit(1)
    sugg=json.load(open(suggestions_path))
    fs=sugg['fileSuggestions'][0]
    items=fs['suggestions']

    access=get_access_token()
    headers={'Authorization':'Bearer '+access}
    fileId=fs.get('fileId')
    # fetch file
    meta_url=f'https://www.googleapis.com/drive/v3/files/{fileId}?fields=id,name,mimeType,modifiedTime,headRevisionId'
    req=urllib.request.Request(meta_url, headers=headers)
    with urllib.request.urlopen(req) as r:
        meta=json.load(r)
    dl_url=f'https://www.googleapis.com/drive/v3/files/{fileId}?alt=media'
    req=urllib.request.Request(dl_url, headers=headers)
    with urllib.request.urlopen(req) as r:
        content=r.read().decode('utf-8')
    lines=content.splitlines()

    backups_dir='outputs/todolist-md/backups'
    os.makedirs(backups_dir, exist_ok=True)
    bk_name=f"{backups_dir}/{fileId}.{datetime.datetime.utcnow().strftime('%Y%m%dT%H%M%SZ')}.md"
    open(bk_name,'w').write(content)
    print('Backup written to',bk_name)

    new_lines=list(lines)
    offset=0
    inserted=0
    records=[]
    for it in items:
        orig=it.get('original')
        suggested=it.get('suggested_text','')
        # extract title from suggested (strip leading - [ ] )
        title=re.sub(r'^\- \[ \]\s*','', suggested).split('\n')[0]
        sim=SequenceMatcher(None, normalize(orig), normalize(title)).ratio()
        actionable=is_actionable(title)
        decision='insert'
        if sim>=args.threshold:
            decision='skip_similar'
        elif not actionable:
            decision='skip_not_actionable'
        records.append({'original':orig,'title':title,'similarity':sim,'actionable':actionable,'decision':decision})
        if decision!='insert':
            continue
        # find best match line index
        best_idx=None; best_score=0
        for idx,ln in enumerate(lines):
            score=SequenceMatcher(None, normalize(orig), normalize(ln)).ratio()
            if score>best_score:
                best_score=score; best_idx=idx
        if best_idx is None or best_score<0.2:
            continue
        human_note=f"  > <!-- bot: note --> {title}"
        sub_id='sub:'+uuid.uuid4().hex[:8]
        marker_obj={'id':sub_id,'original_line':orig,'title':title,'estimate_hours':None,'assignee':None,'createdAtUtc':datetime.datetime.utcnow().isoformat()+'Z','source_model':sugg.get('schema','gpt-5-mini'),'status':'suggested'}
        machine_comment='<!-- bot: subtask '+json.dumps(marker_obj,ensure_ascii=False)+' -->'
        idx=best_idx+1+offset
        window=new_lines[max(0,idx-2):min(len(new_lines),idx+3)]
        if human_note.strip() in '\n'.join(window) or machine_comment in '\n'.join(window):
            continue
        new_lines.insert(idx,human_note)
        new_lines.insert(idx+1,machine_comment)
        offset+=2
        inserted+=1

    # update last_review
    last_line=f'<!-- bot: last_review --> {datetime.datetime.utcnow().isoformat()}Z model=gpt-5-mini hash={sugg.get("generatedAtUtc","")}'
    if new_lines and new_lines[0].startswith('<!-- bot: last_review'):
        new_lines[0]=last_line
    else:
        new_lines.insert(0,last_line)

    new_content='\n'.join(new_lines)+('\n' if content.endswith('\n') else '')
    if args.apply:
        req=urllib.request.Request(f'https://www.googleapis.com/upload/drive/v3/files/{fileId}?uploadType=media&supportsAllDrives=true', data=new_content.encode('utf-8'), method='PATCH')
        req.add_header('Authorization','Bearer '+access)
        req.add_header('Content-Type', meta.get('mimeType','text/markdown'))
        with urllib.request.urlopen(req) as r:
            resp=r.read().decode('utf-8')
        print('Applied, inserted',inserted)
    else:
        print('Dry-run would insert',inserted)
    # write records
    os.makedirs('outputs/todolist-md/filter_records', exist_ok=True)
    recfn=f"outputs/todolist-md/filter_records/{fileId}.{datetime.datetime.utcnow().strftime('%Y%m%dT%H%M%SZ')}.json"
    json.dump(records, open(recfn,'w'), indent=2, ensure_ascii=False)
    print('Records ->',recfn)

if __name__=='__main__':
    main()
